### DISCUSSION TITLE
No such column error on existing column.

### DISCUSSION THREAD

**Discussion by muddi900** [simolus3/drift#2671] [Q&A] [answered]
I am getting this error on insert

```zsh
] Unhandled Exception: SqliteException(1): while preparing statement, no such column: message_index, SQL logic error (code 1)
E/flutter ( 7727):   Causing statement: INSERT INTO "chats" ("id", "last_access_time") VALUES (?, ?)
```

The error is being caused by a trigger:

```sql
CREATE TRIGGER IF NOT EXISTS insert_system_message
AFTER INSERT ON chats
BEGIN
    INSERT INTO messages (role,content,messages.message_index,"chat_id")
    VALUES('system' ,'You are a helpful chat bot that limits your responses to 300 characters.' ,0 , NEW.id);
END;
```

However the `messages` has the `messageIndex` table.

`db.dart`
```dart
class Messages extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get role => text()();
  TextColumn get content => text()();
  IntColumn get messageIndex => integer().nullable()();
  TextColumn get chatId => text().references(Chats, #id)();
  IntColumn get previousMessage =>
      integer().nullable().references(Messages, #id)();
}
```

`db.g.dart`

```dart
class Message extends DataClass implements Insertable<Message> {
  final int id;
  final String role;
  final String content;
  final int? messageIndex;
  final String chatId;
  final int? previousMessage;
  const Message(
      {required this.id,
      required this.role,
      required this.content,
      this.messageIndex,
      required this.chatId,
      this.previousMessage});
  @override
  Map<String, Expression> toColumns(bool nullToAbsent) {
    final map = <String, Expression>{};
    map['id'] = Variable<int>(id);
    map['role'] = Variable<String>(role);
    map['content'] = Variable<String>(content);
    if (!nullToAbsent || messageIndex != null) {
      map['message_index'] = Variable<int>(messageIndex);
    }
    map['chat_id'] = Variable<String>(chatId);
    if (!nullToAbsent || previousMessage != null) {
      map['previous_message'] = Variable<int>(previousMessage);
    }
    return map;
  }
```

**Comment by simolus3** [comment]
Has `messageIndex` been added in a later version of the database (e.g., could this be a migration issue)? You can check that the actual database schema matches what drift is expecting [like this](https://drift.simonbinder.eu/docs/migrations/#verifying-a-database-schema-at-runtime).

**Reply by muddi900** [reply]
This is the my migrationStrategy

```
@override
  MigrationStrategy get migration => MigrationStrategy(
        onCreate: (m) => m.createAll(),
      );
```

**Reply by simolus3** [reply]
Did you add the `messageIndex` column after an earlier version of the app and database has already been installed on the device? You'd have to write a schema migration with `onUpgrade` for that.

**Reply by muddi900** [reply]
I am only working off the emulator and I keep deleting the app every time. Last time I comment out the migration strategy override, but the error persists.

**Reply by muddi900** [reply]
drift_db_viewer shows that the column exists on the table:

![image](https://github.com/simolus3/drift/assets/37051110/ba91f389-84fc-4d5b-8f93-5d7124d9450c)

**Reply by muddi900** [reply]
Also no exception is thrown by the verification.

**Reply by simolus3** [reply]
Thanks for following up.

Something I've only realized now: The `INSERT INTO messages (role,content,messages.message_index,"chat_id")
` looks a bit weird, it should just be `message_index` instead of  `messages.message_index`.
`drift_dev` should have caught that error, do you see any warnings about that when running the build? Could you also post the generated source code for the trigger (the definition of the field `insertSystemMessage` in the database class).

**Reply by muddi900** [reply]
Drift_dev did indeed catch it. But the error persists even after fixing that.

**Reply by simolus3** [reply]
Did you delete the database after fixing it to recreate the trigger? I'm still interested in the trigger's definition as well, so that I can try to reproduce the issue (we rewrite parts of the SQL in generated code, which could also have introduced the bug).

**Reply by muddi900** [reply]
I uninstall the app from the Android Emulator, which should delete the db as well.

```sql
CREATE TRIGGER IF NOT EXISTS insert_system_message
AFTER INSERT ON chats
BEGIN
    INSERT INTO messages (role,content,message_index,"chat_id")
    VALUES('system','You are a helpful chat bot that limits your responses to 300 characters.',0,NEW.id);
END;
```

This is the current def. I deleted the generated file and ran build_runner again.

```bash
[INFO] Generating build script completed, took 686ms
[INFO] Reading cached asset graph completed, took 499ms
[INFO] Checking for updates since last build completed, took 3.3s
[WARNING] No actions completed for 17.8s, waiting on:
  - drift_dev on lib/models/db.expr.temp.dart
  - drift_dev on lib/models/db.dart
  - drift_dev on lib/widgets/image_history.dart
  - drift_dev on lib/screens/chat.dart
  - drift_dev on lib/widgets/image_widget.dart
  .. and 8 more
```

**Reply by muddi900** [reply]
Is there a way I can access the sqlite file using sqlite shell? Because this is baffling.

**Reply by simolus3** [reply]
It really is. If you're using a `NativeDatabase`, you could print the path to the database and pull it via adb. Once you have access to the file, you can use the sqlite3 CLI tools locally.

**Reply by muddi900** [reply, marked as answer]
I figured it out. The PRAGMA output shows that the field is nullable

```
cid         name        type        notnull     dflt_value  pk
----------  ----------  ----------  ----------  ----------  ----------
0           id          INTEGER     1                       1
1           role        TEXT        1                       0
2           content     TEXT        1                       0
3           message_in  INTEGER     0                       0
4           chat_id     TEXT        1                       0
5           previous_m  INTEGER     0                       0
```

Apparently initiating a nullable field in an insert statement throws this error. This true for running a purely sql runtime.

I am updating the field with another trigger.
